mbox: Add NXP MU as a MBOX device

yclistanford opened the pull request:
Add a MBOX driver wrapper around the NXP Freescale MU, simular to the existing wrapper around the NXP S32 MRU. This allows Zephyr IPC to work based on the Freescale MU, on a number of NXP boards.

This is for the following issue:
Support IPC over NXP Freescale message unit #61906

yclistanford committed:
mbox: Add NXP MU as a MBOX device

Add a MBOX driver wrapper around the NXP MU, simular to
the existing wrapper around the NXP S32 MRU. This allows Zephyr IPC
to work based on the MU, on a number of NXP boards.

Also update the SHA of NXP HAL to enable the Kconfig for this driver.

Signed-off-by: Yicheng Li <yichengli@google.com>

decsny commented:
which platform(s) in zephyr are you enabling this for

yclistanford commented:
> which platform(s) in zephyr are you enabling this for

Hi, I am using this for boards/arm/mimxrt595_evk/. The NXP RT595 and a number of other RT boards have the Freescale message unit. The RT595 has a Cortex-M33 core and a DSP. My project uses a device tree overlay out-of-tree to specify the message unit on both sides.

carlocaione commented on drivers/mbox/Kconfig.nxp_fsl:
```diff
@@ -0,0 +1,8 @@
+
+config MBOX_NXP_FSL_MU
+	bool "NXP Freescale Message Unit (MR) driver"
+	default n
```
This should be:
```
default y
depends on DT_HAS_...
```
Please, check how this is done for the other MBOX drivers.

yclistanford replied:
Done

carlocaione commented on drivers/mbox/mbox_nxp_fsl_mu.c:
```diff
@@ -0,0 +1,202 @@
+		}								\
+	}
+
+#if DT_NODE_HAS_STATUS(MU_NODE(0), okay)
+MU_INIT_IRQ_FUNC(0);
+MU_INSTANCE_DEFINE(0);
+MU_IRQ_HANDLER(0);
+#endif
+
+#if DT_NODE_HAS_STATUS(MU_NODE(1), okay)
+MU_INIT_IRQ_FUNC(1);
+MU_INSTANCE_DEFINE(1);
+MU_IRQ_HANDLER(1);
+#endif
```
Why not using something more generic like `DT_INST_FOREACH_STATUS_OKAY`?

yclistanford replied:
Done

dbaluta commented:
@yclistanford  I think we should remove Freescale from naming. `NXP MU...` works perfectly fine.

dbaluta commented on drivers/mbox/CMakeLists.txt:
```diff
@@ -7,4 +7,5 @@ zephyr_library()
 zephyr_library_sources_ifdef(CONFIG_MBOX_NRFX_IPC   mbox_nrfx_ipc.c)
 zephyr_library_sources_ifdef(CONFIG_MBOX_NXP_S32_MRU   mbox_nxp_s32_mru.c)
+zephyr_library_sources_ifdef(CONFIG_MBOX_NXP_FSL_MU   mbox_nxp_fsl_mu.c)
 zephyr_library_sources_ifdef(CONFIG_MBOX_ANDES_PLIC_SW   mbox_andes_plic_sw.c)
```
Here should be `mbox_nxp_imx_mu` instead of `mbox_nxp_fsl`...

yclistanford replied:
To clarify, would you like the .c file to be named `mbox_nxp_imx_mu.c`, and the Kconfig to be named `CONFIG_MBOX_NXP_IMX_MU`, and all the driver functions to be prefixed with `nxp_imx_mu_`?

dbaluta replied:
Yes.

dbaluta commented:
I'm new to zephyr but can anyone clarify what's the difference between IPM and MBOX?

https://docs.zephyrproject.org/latest/hardware/peripherals/ipm.html

Looks like IPM supports NXP MU.

@yclistanford also, do you plan to add support for doorbell with your nxp mbox driver?

carlocaione commented:
> I'm new to zephyr but can anyone clarify what's the difference between IPM and MBOX?

IPM is to be considered legacy and MBOX is the new API with better DT support and multi-channels support.

decsny commented:
> Hi, I am using this for boards/arm/mimxrt595_evk/. The NXP RT595 and a number of other RT boards have the Freescale message unit. The RT595 has a Cortex-M33 core and a DSP. My project uses a device tree overlay out-of-tree to specify the message unit on both sides.

Thanks, can you add the DT node at least to the file dts/arm/nxp/nxp_rt5xx_common.dtsi ? And maybe also something to the board evk dts as well? Otherwise this driver has no platform example of it's use in the zephyr tree.

iuliana-prodan commented:
@yclistanford what is the exact use case for this?
I understand you want the M33 and DSP cores to communicate on RT595, right?
Can't this be done with [ipm_imx](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ipm/ipm_imx.c)?

Why do you need to enable both general purpose interrupts and RX?

carlocaione commented:
> @yclistanford what is the exact use case for this? I understand you want the M33 and DSP cores to communicate on RT595, right? Can't this be done with [ipm_imx](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ipm/ipm_imx.c)?

Because IPM is legacy and we really should be stopping using it given that it doesn't support IPC service and it doesn't have a proper DT support. On the contrary I really welcome this MBOX driver.

iuliana-prodan commented on drivers/mbox/mbox_nxp_fsl_mu.c:
```diff
@@ -0,0 +1,202 @@
+
+	/* Signalling mode. */
+	if (msg == NULL) {
+		if (kStatus_Success != MU_TriggerInterrupts(
```
`MU_TriggerInterrupts()` already checks if the previous interrupt has been processed - https://github.com/nxp-mcuxpresso/mcux-sdk/blob/main/drivers/mu/fsl_mu.c#L167

yclistanford replied:
I understand that `MU_TriggerInterrupts()` will check and return an error if the previous interrupt has not been processed. I wanted to notify the caller of mbox_send() if that's the case. Would you like me to just do
`return MU_TriggerInterrupts(cfg->base, kMU_GenInt0InterruptTrigger);` ? That will cast the status_t to an int.

iuliana-prodan replied:
Yes, I would call it directly, no need for an extra check

iuliana-prodan commented on drivers/mbox/mbox_nxp_fsl_mu.c:
```diff
@@ -0,0 +1,202 @@
+
+	if (enable) {
+		MU_EnableInterrupts(cfg->base,
+			(kMU_GenInt0InterruptEnable | kMU_Rx0FullInterruptEnable));
```
Why do you enable/disable only general purpose interrupt 0 and RX0?

yclistanford replied:
General purpose interrupt 0 and RX0 are the minimum we need to support running Zephyr IPC service over this hardware. Doesn't have to be index 0, could be any other general purpose interrupt and RX. Enabling the other RXs can let this MBOX device send more data at once, I am ok with either just RX0 or more. Enabling other general interrupts is not helpful I think, since the [IPC service](https://docs.zephyrproject.org/latest/services/ipc/ipc_service/ipc_service.html) only needs one mechanism to notify the other side.

iuliana-prodan replied:
I understand is working. This is great! :)
But, there is no consistency here.
You defined the MU_MBOX_SIZE and MU_MAX_MBOX_PER_CHAN ([here](https://github.com/zephyrproject-rtos/zephyr/pull/61910/files#diff-d617dd790b24bcbd937ef8d7508a31367c45869e809006a2f89f2fcf6d3f181dR18)) and based on these you should enable the interrupts.
So, in this case, you should enable all interrupts.

See an example here  - the legacy [IPM_IMX](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ipm/ipm_imx.c#L253) which uses the same MU.

iuliana-prodan commented:
> > @yclistanford what is the exact use case for this? I understand you want the M33 and DSP cores to communicate on RT595, right? Can't this be done with [ipm_imx](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ipm/ipm_imx.c)?
> 
> Because IPM is legacy and we really should be stopping using it given that it doesn't support IPC service and it doesn't have a proper DT support. On the contrary I really welcome this MBOX driver.

If one adds new code just because something is legacy, than the legacy code should be modify to use the new driver (like the IPC samples). Adding a driver just to be there, I don't see it as a good practice - is just more work to be maintained.

@yclistanford can you please, add a use-case, a sample or a node in the dts?

carlocaione commented:
> If one adds new code just because something is legacy, than the legacy code should be modify to use the new driver (like the IPC samples). Adding a driver just to be there, I don't see it as a good practice - is just more work to be maintained.

Agree, but this is a good first step in that direction and I do not see a valid reason to discourage this. The IPC stuff can be modified / ported if needed, or simply a new proper IPC service implementation can be added in parallel, I don't see any problem with that.

I asked several times the NXP folks to take a look at the new IPC service framework because the old RPMSg service is really something that we should deprecate but nothing came out of that. So I'm happy to see some movement in that direction, even if this is coming from google folks.

dbaluta commented:
> I asked several times the NXP folks to take a look at the new IPC service framework because the old RPMSg service is really something that we should deprecate but nothing came out of that. So I'm happy to see some movement in that direction, even if this is coming from google folks.

@carlocaione I've recently landed into Zephyr stuff. Can you please share with us the PR for the new IPC service framework? I would gladly offer some review.

carlocaione commented:
> @carlocaione I've recently landed into Zephyr stuff. Can you please share with us the PR for the new IPC service framework? I would gladly offer some review.

It is not that new nowadays so there are several PRs and a couple of backends already but I can give you some pointers:

- https://docs.zephyrproject.org/latest/services/ipc/ipc_service/ipc_service.html (general doc, pretty basic)
- https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c (this is the `static VRINGs RPMSg` backend that is the equivalent of the old RPMSg service implemented on top of the generic IPC service. It uses OpenAMP)
- https://github.com/zephyrproject-rtos/zephyr/pull/40365 (the PR introducing the RPMSg backend)
- https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/ipc/ipc_service/static_vrings (the sample showing how to actually use it)

The whole point of the IPC service is that you can take the sample and you can change the backend (moving from RPMSg to something maybe custom) without changing one single line of code. From the user prospective you have a common and standard set of functions to perform IPC without having to worry what kind of transport protocol, backend or code is implemented behind the curtains to move the data from one endpoint to the other.

As I specified before we currently have two backends, one being RPMSg implemented using OpenAMP. With this backend you can fully specify your IPC instances and endpoints in the DT, the only constrain you have is that the backend is assuming that you are using MBOXes for signalling (in theory IPM can be used but someone must add support for that).

Objectively this could be better documented and I take responsibility for this.

yclistanford commented:
@dbaluta

> @yclistanford I think we should remove Freescale from naming. `NXP MU...` works perfectly fine.

Done

> @yclistanford also, do you plan to add support for doorbell with your nxp mbox driver?

I am not sure what you mean by doorbell?

yclistanford commented:
@decsny and also @iuliana-prodan 

> Thanks, can you add a DT node at least to the file dts/arm/nxp/nxp_rt5xx_common.dtsi ? Otherwise this driver has no platform example in the zephyr tree.

I added the DT node to `dts/arm/nxp/nxp_rt5xx_common.dtsi`, and added the example for board `mimxrt595_evk_cm33` to `samples/drivers/mbox`, with 2 caveats:
1. Building this sample requires adding a line `include_driver_ifdef(CONFIG_MBOX_NXP_FSL_MU  mu  driver_mu)` to the `modules/hal/nxp/` repository, which I haven't uploaded yet.
2. `BOARD_REMOTE` in `samples/drivers/mbox/CMakeLists.txt` is supposed to be the DSP on that board, but Zephyr support for that DSP is pending this PR from my collegue: https://github.com/zephyrproject-rtos/zephyr/pull/61356. So I put BOARD_REMOTE=nrf5340dk_nrf5340_cpunet in there just to make sure the sample builds. We should address this before we merge this PR.

With the above 2 caveats I tested I can build `samples/drivers/mbox` with `--board mimxrt595_evk_cm33`

yclistanford commented:
> @yclistanford what is the exact use case for this? I understand you want the M33 and DSP cores to communicate on RT595, right? Can't this be done with [ipm_imx](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ipm/ipm_imx.c)?

The use case is to send a buffer of > 1000 bytes from the M33 to the DSP, without copying. Looks like I need to use [this API from Zephyr IPC service](https://docs.zephyrproject.org/latest/services/ipc/ipc_service/ipc_service.html#c.ipc_service_send_nocopy).

Admittedly, one can allocate/manage buffers in shared memory between the M33 and the DSP, and send the address of the buffer through IPM, but that adds complexity and fragility to M33 software, and I feel we should really utilize Zephyr IPC service.

> Why do you need to enable both general purpose interrupts and RX?

The Zephyr [MBOX model](https://docs.zephyrproject.org/latest/hardware/peripherals/mbox.html) says it's supposed to support both signaling mode and data transfer mode. In signaling mode, no bytes are sent, so the general purpose interrupt seems more natural than RX. In data transfer mode the RX_full interrupt is needed.

The [icmsg service backend](https://github.com/zephyrproject-rtos/zephyr/blob/5689916a70ad5c363bd7d5511b772ada90067d51/subsys/ipc/ipc_service/lib/icmsg.c#L453) uses the signaling mode to notify the remote when sending.

yclistanford commented:
> > > @yclistanford what is the exact use case for this? I understand you want the M33 and DSP cores to communicate on RT595, right? Can't this be done with [ipm_imx](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ipm/ipm_imx.c)?
> > 
> > 
> > Because IPM is legacy and we really should be stopping using it given that it doesn't support IPC service and it doesn't have a proper DT support. On the contrary I really welcome this MBOX driver.
> 
> If one adds new code just because something is legacy, than the legacy code should be modify to use the new driver (like the IPC samples). Adding a driver just to be there, I don't see it as a good practice - is just more work to be maintained.
> 
> @yclistanford can you, please, add a use-case, a sample or a node in the dts?

@iuliana-prodan I added the DT node and sample now. Please see my reply to @decsny

yclistanford commented:
The twister build `west build -p -b mimxrt595_evk_cm33 -T samples/drivers/mbox/sample.drivers.mbox` will be failing until we update the `modules/hal/nxp` repository to include the MU driver upon `CONFIG_MBOX_NXP_FSL_MU`

carlocaione commented:
> The twister build `west build -p -b mimxrt595_evk_cm33 -T samples/drivers/mbox/sample.drivers.mbox` will be failing until we update the `modules/hal/nxp` repository to include the MU driver upon `CONFIG_MBOX_NXP_FSL_MU`

@yclistanford do you plan to do that?

dbaluta commented:
> I am not sure what you mean by doorbell?

Doorbell is just when you send an interrupt without an actual data. NXP MU has functionality for this. I think with MBOX model this is called signalling mode.

dbaluta commented on drivers/mbox/mbox_nxp_fsl_mu.c:
```diff
@@ -0,0 +1,202 @@
+#define DT_DRV_COMPAT nxp_fsl_mu
+
+#define MU_MAX_CHANNELS 2
+
```
MU has 4 max channels. Why use 2 here?

iuliana-prodan commented on dts/bindings/mbox/nxp,fsl-mu.yaml:
```diff
@@ -0,0 +1,29 @@
+      set rx-channels of mruA as follows:
+
+        mruA {
+          rx-channels = <2>;
```
There are 4 RX channels

decsny commented:
> 1. Building this sample requires adding a line `include_driver_ifdef(CONFIG_MBOX_NXP_FSL_MU  mu  driver_mu)` to the `modules/hal/nxp/` repository, which I haven't uploaded yet.

please do that.

yclistanford commented:
> > The twister build `west build -p -b mimxrt595_evk_cm33 -T samples/drivers/mbox/sample.drivers.mbox` will be failing until we update the `modules/hal/nxp` repository to include the MU driver upon `CONFIG_MBOX_NXP_FSL_MU`
> 
> @yclistanford do you plan to do that?

@carlocaione @decsny 

Done in https://github.com/zephyrproject-rtos/hal_nxp/pull/269

iuliana-prodan commented on drivers/mbox/mbox_nxp_imx_mu.c:
```diff
@@ -0,0 +1,201 @@
+	}
+
+	if (enable) {
+		if (data->cb[channel] == NULL) {
+			LOG_WRN("Enabling channel without a registered callback");
+		}
```
@carlocaione is the `.register_callback` mandatory? 
In my opinion is not.

Also, the `mbox_register_callback` might be called after `mbox_set_enabled` - https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/ipc/ipc_service/lib/icmsg.c#L31
And, in this driver, the` data->cb` is checked before used - https://github.com/zephyrproject-rtos/zephyr/blob/6224b604d34ca44e6cc9f36d86206e9b2faf59b7/drivers/mbox/mbox_nxp_imx_mu.c#L189

carlocaione commented on drivers/mbox/mbox_nxp_imx_mu.c:
```diff
@@ -0,0 +1,183 @@
+		return -EMSGSIZE;
+	}
+
+	/* memcpy to avoid issues when msg->data is not word-aligned. */
+	memcpy(&data, msg->data, sizeof(uint32_t));
```
This is a bit odd. First of all you are neglecting `msg->size` (you did a lot of error checking before but never actually used it) also there is no guarantee that `data` will be aligned as well (sure, this is extremely probable but you still have to trust your compiler).

carlocaione reviewed (commented):
Just a comment.

carlocaione commented:
@yclistanford also CI is failing.
